File: Data Structures & Algorithms/non-cyclical-number/submission-1.py
class Solution:
    def isHappy(self, n: int) -> bool:
        s = set()
        while n:
 
            x = 0
            while n:
                x += (n % 10)**2
                n = n // 10
            n = x
            print(n)
            if n == 1:
                return True
            if n in s:
                return False
            s.add(n)

File: Data Structures & Algorithms/clone-graph/submission-1.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    unordered_map<Node*, Node*> umap;
    Node* cloneGraph(Node* node) {
        if(node == nullptr) return node;
        if(umap.count(node)) return umap[node];
        umap[node] = new Node(node->val);
        for(auto nei: node->neighbors) {
            umap[node]->neighbors.push_back(cloneGraph(nei));
        }
    
        return umap[node];
    }
};
